### ISSUE TITLE
CompareInfo.IsSuffix sometimes incorrectly splits collation elements and returns 'true'

### ISSUE THREAD

**Issue by GrabYourPitchforks** [dotnet/runtime#43897]
Repro (.NET 5 on ICU):

```cs
var ci = CultureInfo.InvariantCulture.CompareInfo;
int idx = ci.IndexOf("\u0600x", "x", CompareOptions.None);
bool isSuffix = ci.IsSuffix("\u0600x", "x", CompareOptions.None);
Console.WriteLine($"idx = {idx}, isSuffix = {isSuffix}");
```

This sample prints __`idx = -1, isSuffix = True`__. These return values are inconsistent with respect to each other. If _IsSuffix_ returns true, then _IndexOf_ must return a non-negative value. It's the contrapositive of the code comment at:

https://github.com/dotnet/runtime/blob/eb187d1250cf65c16442c207ff97da65690afae5/src/libraries/System.Private.CoreLib/src/System/String.Searching.cs#L270-L273

**Comment by tarekgh** [comment]
https://github.com/dotnet/runtime/issues/40803

**Comment by GrabYourPitchforks** [comment]
Note: This sample uses `CompareOptions.None`. The other linked issue considers non-default `CompareOptions` arguments.

**Comment by tarekgh** [comment]
> Note: This sample uses CompareOptions.None. The other linked issue considers non-default CompareOptions arguments.

That is right. but the root cause the same which is coming from the optimization of this case. it is ok to keep both issues or close one and add the extra information to the other.

**Comment by adamsitnik** [comment]
#40803 is not related to the "optimizaiton", this issue most probably is. I think it's best to keep both of them open until we fix them.

**Comment by adamsitnik** [comment]
I've confirmed that `usearch_first` returns `-1` for this particular case, while the first element returned by `ucol_next(pSourceIterator)` returns `UCOL_IGNORABLE` and hence the "optimized" version of the code assumes that the first character (`\u0600`) can be ignored and later returns true

![obraz](https://user-images.githubusercontent.com/6011991/97729552-210f3b00-1ad3-11eb-8ac4-401e1ca38dbb.png)

![obraz](https://user-images.githubusercontent.com/6011991/97729279-d988af00-1ad2-11eb-8ac9-2bff88c34212.png)

I've also tried it with .NET Core 3.1 on Windows to see what NLS has to say:

![obraz](https://user-images.githubusercontent.com/6011991/97730200-dd690100-1ad3-11eb-9976-1c1a023457f5.png)

And for NLS it returns `1` and `true`.

Which makes be confused. 

@GrabYourPitchforks @tarekgh In a perfect world, where there are no Unicode bugs, what should be the expected result and why?

**Comment by adamsitnik** [comment]
@GrabYourPitchforks I've raised a question in the ICU bug tracker: https://unicode-org.atlassian.net/browse/ICU-21375

**Comment by tarekgh** [comment]
note that I expect this will happen with other codepoints (e.g. `\u0601` too). I have some theory around these but let's wait if anyone officially reply the bug.

**Comment by adamsitnik** [comment]
My question got [answered by one of the ICU members](https://unicode-org.atlassian.net/browse/ICU-21375?focusedCommentId=159294):

> This looks like it's working as intended. U+0600 is one of many characters that map to a completely ignorable collation element: https://www.unicode.org/charts/collation/ --> https://www.unicode.org/charts/collation/chart_Ignored.html

So `StartsWith` works as expected?

> I don't know why `usearch_first()` returns -1 for you. The StringSearch/SearchIterator classes have several options for how to deal with ignorables and combining marks etc. at the edges of a match. I suspect that there is optional behavior on top of collation element comparison responsible for rejecting the match.

I am able to get `-1` for a small repro without any custom settings. @tarekgh @GrabYourPitchforks should I file a bug for `usearch_first` then?

**Comment by tarekgh** [comment]
@adamsitnik I don't think this is a bug in ICU. I am seeing `U+0600` is one of the characters that can be combined with the following characters which change the search operation https://www.unicode.org/Public/UCD/latest/ucd/auxiliary/GraphemeBreakProperty.txt. Looks by design to ICU. This is similar to `\r\n` case. That is my understanding but I cannot claim I am expert in that :-)

**Comment by adamsitnik** [comment]
@tarekgh so `ucol_next(pSourceIterator)` returns `UCOL_IGNORABLE` for `U+0600` because as a single character it's ignorable, but when combined with the next character it's not ignorable anymore? 

It would explain why ICU's `usearch_search` is not using the `ucol_next` API that works on `int32_t` collation elements, but uses a `int64_t `-based API instead:

https://github.com/adamsitnik/icu/blob/3c3040eaf8a60919ae9ffd3d3804abd59ed22316/icu4c/source/i18n/usearch.cpp#L3932

https://github.com/unicode-org/icu/blob/f744742e036441cbfd54b535e722dc3801ff0f63/icu4c/source/i18n/usearch.cpp#L3766

the problem is that this API is private and we can't use it ;/

**Comment by tarekgh** [comment]
> but when combined with the next character it's not ignorable anymore?

Right. when combining two or more characters into a grapheme, this will make the whole grapheme treated as one unit and will have none ignorable weight. 

@adamsitnik can we measure if we removed this optimization how much we lose? as you see I have optimized the search operation which should help in such scenarios too. I am afraid whatever we do here we'll run into more problems that we never encountered before.

**Comment by adamsitnik** [comment]
> can we measure if we removed this optimization how much we lose? 

@tarekgh we can, but I don't think it's worth it. The problem with the initial implementation was searching the entire text for a pattern for cases where the given text was not starting|ending with the given pattern. Your optimization has helped with reducing the cost of initialization of internal ICU structures, but for long strings that don't start|end with given prefix|sufix we would still do the very time-consuming comparison. 

Example: 
```cs
new string('a', 10_000_00).StartsWith("b")`
``` 

Would result in:

- one cache access instead of 1 expensive initialization (great 👍 )
- 10_000_00 comparisons instead of one (not good)

What other options do we have? 

Should we propose adding a new `StartsWith` and `EndsWith` methods to ICU API? But it would then work only for those of our customers who have the latest version installed ;/

I wonder how Windows solved this problem as most probably they have faced the same issue. I've noticed that Microsoft has its own fork of ICU: https://github.com/microsoft/icu Have we considered shipping our own copy of ICU with .NET? (I am just thinking out loud, I am not a fan of forking other projects)

**Comment by tarekgh** [comment]
> I wonder how Windows solved this problem as most probably they have faced the same issue

I don't think this affect Windows at all. Windows either provide NLS APIs which is not using ICU anyway. Or providing ICU as it is without any additions. 

> Have we considered shipping our own copy of ICU with .NET?

That is exactly what we are trying to avoid for years :-) as soon as we own it, we'll run into many issues we are trying to avoid (e.g. servicing it). That is why we push the issue to the user and the OS which maintain it.

**Comment by tarekgh** [comment]
Is it possible we can look at `ucol_nextSortKeyPart()`. I think this may help us implementing StartWith/EndsWith more reliably. but need some investigation when we have a time. The idea would be comparing the sort keys instead of comparing the code units.

**Comment by adamsitnik** [comment]
I have wrongly assumed that at some point Windows is going to change the internals of NLS API to be implemented by using ICU inside (and hit the point they need to implement startswith|endswith).

One more idea is to use [BreakIterators](http://userguide.icu-project.org/boundaryanalysis)

**Comment by tarekgh** [comment]
I have looked deeply on this issue, and I am seeing the underlying problem is coming from ICU and not from our code. 

When using ICU and comparing the two strings `x` and `\u0600`, it returns the result indicating the two strings are equal which is correct. The reason is the character `\u0600' is ignorable character according to [Unicode Standard](https://www.unicode.org/charts/collation/chart_Ignored.html). But when try to search for `x` in the string `\u0600x` using ICU APIs like `usearch_search` or `usearch_last`, it returns `-1` which is wrong.

I have reported the issue to ICU through the [ticket](https://unicode-org.atlassian.net/browse/ICU-21698). I mentioned the details in the ticket and the possible root cause of this problem.
